src/status_im/chat/commands/core.cljs

```diff
@@ -0,0 +1,139 @@
+(ns status-im.chat.commands.core
+  (:require-macros [status-im.utils.views :refer [defview letsubs]])
+  (:require [re-frame.core :as re-frame]
+            [status-im.ui.components.react :as react]
+            [status-im.ui.components.icons.vector-icons :as vector-icons]
+            [status-im.ui.components.colors :as colors]
+            [status-im.i18n :as i18n]
+            [status-im.chat.styles.message.message :as style]))
+
+(defprotocol Command
+  "Protocol for defining command message behaviour"
+  (id [this] "Identifier of the command, used to look-up command display name as well")
+  (scope [this]
+    "Scope of the command, defined as set of values representing contexts
+     where in which command is available, together with `id` it forms unique
+     identifier for each command.
+     Available values for the set are: 
+     `id-of-the-any-chat` - command if available only for the specified chat
+     `:personal-chats` - command is available for any personal 1-1 chat
+     `:group-chats` - command is available for any group chat
+     `:public-chats` - command is available for any public chat
+     `:requested` - command is available only when there is an outstanding request")
+  (parameters [this]
+    "Ordered sequence of command parameter templates, where each parameter 
+     is defined as map consisting of mandatory `:id`, `:title` and `:type` keys, 
+     and optional `:suggestions` field.
+     When used, `:suggestions` containes reference to any generic helper component
+     rendering suggestions for the argument (input code will handle when and where
+     to render it)")
+  (validate [this parameters cofx]
```

Comment:
When does this validation fires? Do you have an example on when this is useful?

Reply:
The validation fires as a first thing after you hit `send` button, before message is actually sent (saved to app-db/realm, sent over wire, etc.).
I can't think of good example now when we don't do any validation for `/send` (instead delegate that to the wallet module), maybe `/request` validating if the recipient is actually present in chat.
I will add example of the `/request` command definition with validation as well.

Reply:
Isn't that redundant with the parameter `type` ? We could have some specific type that would constraints to a specific list of values for instances (e.g. visible-tokens)

Reply:
Good point, we can have generic validation tied to the each command argument based on its type (very generic like `:text` or `:number`, or more specific like `token`).
The problem is that sometimes you want to do something which requires broader validation with more context, for example `/send` command (hypothetical one not going through the wallet) can validate if you have enough balance of particular asset, etc, that's why I think one method taking `parameters` and whole `cofx` context is still helpful in some cases.
Of course, you can always just leave it blank (returning nil) when such validation is not necessary.

Reply:
Make sense. I was trying to see how this could correlate with `extensions` but I guess both don't have to perfectly align.

Reply:
We actually do still have some validation on the bot side for send and request (number formatting, valid and visible token)